compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -1835,6 +1836,56 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
     }
 
+    pub(super) fn collect_unused_for_coerce_return_ty(
+        &self,
+        errors: &[traits::FulfillmentError<'tcx>],
+    ) {
+        for fillment_error in &*errors {
+            let obligation = &fillment_error.obligation;
+            let body_id = obligation.cause.body_id;
+            let span = obligation.cause.span;
+
+            let Some(body_id) = self.tcx.hir().maybe_body_owned_by(
+                self.tcx.typeck_root_def_id(body_id.to_def_id()).expect_local(),
+            ) else {
+                continue;
+            };
+
+            let Some(mut diag) =
+                self.tcx.sess.diagnostic().steal_diagnostic(span, StashKey::MaybeForgetReturn)
+            else {
+                continue;
+            };
+
+            let body = self.tcx.hir().body(body_id);
+            let root_obligation = &fillment_error.root_obligation;
+            if let Some((_, fn_decl, _)) = self.get_fn_decl(body.value.hir_id)
+                && let hir::FnRetTy::Return(_ty) = fn_decl.output
```

Comment:
Are the first two `let` in the `if` here necessary anymore?

Reply:
Also shouldn't be necessary to get the body either.

Reply:
I was trying to avoid `DefaultReturn`, just found out I can use `!fn_sig.output().is_unit()` to check it.

Reply:
Well, also, this should never trigger if the return type is unit, right? That would mean `can_coerce(...)` to unit, which only works when the type is unit, which should never have an ambiguity error 😅

Reply:
yes, for code:

```rust
fn with_closure<F, A, B>(_: F)
where
    F: FnOnce(A, B),
{
}

fn a() {
    // Type of `y` is unconstrained.
    with_closure(|x: u32, y| {}); //~ ERROR E0282
}

fn main() {}
```

this output is not good😆:

```console
 --> ./p/debug.rs:9:27
  |
9 |     with_closure(|x: u32, y| {}); //~ ERROR E0282
  |                           ^
  |
help: consider giving this closure parameter an explicit type
  |
9 |     with_closure(|x: u32, y: /* Type */| {}); //~ ERROR E0282
  |                            ++++++++++++
help: consider adding `return` here to coerce to expected return type
  |
9 |     return with_closure(|x: u32, y| {}); //~ ERROR E0282
  |     ++++++
```

Reply:
This slightly modified code still causes an awkward suggestion:

```rust
fn with_closure<F, A, B>(_: F) -> i32
where
    F: FnOnce(A, B),
{
    0
}

fn a() -> i32 {
    with_closure(|x: u32, y| {});
    0
}

fn main() {}
```

```
error[E0282]: type annotations needed
 --> /home/gh-compiler-errors/test.rs:9:27
  |
9 |     with_closure(|x: u32, y| {});
  |                           ^
  |
help: consider giving this closure parameter an explicit type
  |
9 |     with_closure(|x: u32, y: /* Type */| {});
  |                            ++++++++++++
help: consider adding `return` here to coerce to expected return type
  |
9 |     return with_closure(|x: u32, y| {});
  |     ++++++

error: aborting due to previous error
```

Reply:
The problem here is walking too far up the HIR parent tree, I think. Not sure if there's a good way around it, needs some brainstorming.

Reply:
I was also guessing it walks too far, but actually only 2 calls on the parent from the log, `!fn_sig.output().is_unit()` filter it out.

Reply:
ah, didn't notice the slightly change you add, let me think ...

Reply:
This is a more simple case:
```rust
fn bad(x: bool) -> Result<(), MyError> {
    Err(MyError); //~ ERROR type annotations needed
    Ok(())
}
```
this is weird:
```console
error[E0282]: type annotations needed
 --> ./p/err.rs:5:9
  |
5 |         Err(MyError);
  |         ^^^ cannot infer type of the type parameter `T` declared on the enum `Result`
  |
help: consider specifying the generic arguments
  |
5 |         Err::<T, MyError>(MyError);
  |            ++++++++++++++
help: consider adding `return` here to coerce to expected return type
  |
5 |         return Err(MyError);
  |         ++++++
```

I changed the code to calculate how many `block`s from the `semi` to function body, only emit the suggestion for cond `block_cnt > 1`, not best solution but seems it works🤔